### PR TITLE
Implementation of tagging system

### PR DESCRIPTION
**Pull request by giplt**
Adds a tag index to a datatype similar to adding a page index. Still lacking in this implementation:
- add styling for default theme, @tammoterhark: can you take a look at this?
- implement error notification if new tag already exists in the tagList
- implement tag management page: edit, translate, remove or merge tags. Can be done (for the time being) by editing hypha.xml.
- implement tagcontainer pagetype.

### CODE REVIEW

File: system/core/tags.php [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		
+		// Construct html output
+		$html = '';
+		if (!empty($activeTags) || isUser()) {
+			$html.= '<span class="prefix">' . __('tags') . ': </span>';
+			$html.= '<div id="selectedTags" style="display:inline-block;">';
+			if (!empty($selectedTags)) foreach ($selectedTags as $id => $tag) {
+				$html.= '<div id="tagSel_'.$id.'" class="tag">';
+				$html.= '<span title="'.$tag['description'].'">'.$tag['label'].'</span>';
+				if (isUser()) $html.= ' <span style="color:#8f0000; cursor:pointer;" title="'.__('remove-tag').'" onclick="deselectTag(\''.$id.'\');">⨯</span>';
```
Comment (tammoterhark) [inline system/core/tags.php]: Please do not use inline styling, this can better be tackled in CSS. Better: 

<span class="delete-tag" />

That way the stylist can put here anything (I think)

File: system/languages/en.php [hunk 2]
```diff
@@ -192,6 +192,8 @@
 		"missing-arguments" => "missing arguments",
 		"archive" => "archive",
 		"sure-to-delete" => "Are you sure you want to delete this page",
+		"enter-label" => "Enter label (use a short word)",
```
Comment (dianawi) [inline system/languages/en.php]: I think it is clearer to use the word Tag here instead of Label. Also it is nice to make the communication with users a bit more human, so I would suggest "Please enter a label/tag" and "Enter a short description". The same goed for the Dutch translation.

File: system/core/events.php [hunk 3]
```diff
@@ -53,12 +53,6 @@ function addEventHandler($html) {
 		ob_start();
 ?>
 <script>
-	/*
-		Variable: baseUrl
-		Javascript variable containing the base url from the document location object.
-	*/
-	var postProcessingList = new Array();
```
Comment (matthijskooijman) [inline system/core/events.php]: This removes the `postProcessingList` without comment. Is it unused and therefore removed? Is it replaced by something else? This should probably be a separate commit in any case?

File: system/core/events.php [hunk 4]
```diff
@@ -67,15 +61,18 @@ function addEventHandler($html) {
 		url - the hypha-formatted page identifier, e.g. 'en/hompage/defaultview'. When left empty the current paged is reloaded.
 		cmd - php function to execute, e.g. 'save'
 		arg - argument to pass to the cmd function
+		callback - optional callback function for making ajax calls
 	*/
-	function hypha(url, cmd, arg) {
+	function hypha(url, cmd, arg, callback = null) {
```
Comment (matthijskooijman) [inline system/core/events.php]: Adding support for a callback here should probably be a separate commit (not a separate PR, though).

I also wonder if adding a callback here is clear enough, since now adding the callback argument in a call completely changes the behaviour of this function. I wonder if it would be better to add the cmd and arg arguments to the `ajax` function (or add a flavour of that function with those arguments)?

File: system/core/tags.php [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	// - add styling for default theme
+	// - implement error notification if new tag exists
+	// - implement tag management page: edit, translate, remove or merge tags
+	// - implement tagcontainer pagetype
+	
+	function tagList($pageListNode, $lang) {
+		global $hyphaXml;
+		
+		// Build list of active tags for given page
+		foreach($pageListNode->getElementsByTagName('tag') as $tag) $activeTags[] = $tag->getAttribute('id');
```
Comment (matthijskooijman) [inline system/core/tags.php]: I believe getElementsByTagName returns all descendent tags, not just direct children. Is that intended here? It probably doesn't hurt, though.

File: system/core/tags.php [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+		
+		// Divide list of all tags stored in hypha.xml in a selected tag list and a deselected tag list
+		foreach($hyphaXml->getElementsByTagName('tag') as $tag) {
```
Comment (matthijskooijman) [inline system/core/tags.php]: This also finds *all* tag elements, which I think is not intended here? E.g. it finds the tag definitions, but also the tag references inside the page list. In practice the latter do not have any `language` children elements, so this doesn't break, but it is at least inefficient and prone to breakage in the future.

I would suggest using Xpath, e.g. something like `$hyphaXml->findXPath("/tags/tag")` (not tested, but I think `$hyphaXml` should have a `findXPath()` method, and I'm just guessing about the XML structure that you need to match with this Xpath expression).

File: system/core/tags.php [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+			foreach($tag->getElementsByTagName('language') as $tagLang) {
+				if ($tagLang->getAttribute('id')==$lang) {
+					if (!empty($activeTags) && in_array($tag->getAttribute('id'), $activeTags)) $selectedTags[$tag->getAttribute('id')] = array("label" => $tagLang->getAttribute('label'), "description" => $tagLang->nodeValue);
```
Comment (matthijskooijman) [inline system/core/tags.php]: This seems like an overly long line to me, I'd add some line break(s).

Also, I think that if you assign an empty array to `$activeTags` at the start of the function, you won't have to do the `empty` check here.

File: system/core/tags.php (hunk 6 above)
Comment (matthijskooijman) [reply system/core/tags.php]: I wonder if it would be a good idea to let the xpath query lookup the tags by language (and perhaps id)? That might make the code simpler.

File: system/core/tags.php [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+			}
+		}
+		
+		// Construct html output
+		$html = '';
+		if (!empty($activeTags) || isUser()) {
+			$html.= '<span class="prefix">' . __('tags') . ': </span>';
+			$html.= '<div id="selectedTags" style="display:inline-block;">';
+			if (!empty($selectedTags)) foreach ($selectedTags as $id => $tag) {
+				$html.= '<div id="tagSel_'.$id.'" class="tag">';
```
Comment (matthijskooijman) [inline system/core/tags.php]: This needs `htmlspecialchars` on the `$id`, and also the tag description and label below.

File: system/core/tags.php [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+				}
+			}
+		}
+		
+		// Construct html output
+		$html = '';
+		if (!empty($activeTags) || isUser()) {
+			$html.= '<span class="prefix">' . __('tags') . ': </span>';
+			$html.= '<div id="selectedTags" style="display:inline-block;">';
+			if (!empty($selectedTags)) foreach ($selectedTags as $id => $tag) {
```
Comment (matthijskooijman) [inline system/core/tags.php]: Again, initializing `$selectedTags` to the empty array first removes the need for this if (probably the same for deselected tags).

File: system/core/tags.php (hunk 1 above)
Comment (matthijskooijman) [reply system/core/tags.php]: I would write the onclick part as:

`'onclick="deselectTag(' . htmlspecialchars(json_encode($id)).');"'`

This makes sure that any special characters will not mess up the HTML, or the javascript string literal. Note that his uses json_encode to encode a javascript string, which also generates to surrounding quotes.

File: system/core/tags.php [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+				$html.= '<span title="'.$tag['description'].'">'.$tag['label'].'</span>';
+				if (isUser()) $html.= ' <span style="color:#8f0000; cursor:pointer;" title="'.__('remove-tag').'" onclick="deselectTag(\''.$id.'\');">⨯</span>';
+				$html.= '</div>';
+			}
+			$html.= '</div>';
+			if ((isUser() && !empty($deselectedTags)) || isAdmin()) {
+				$html.= '<select id="remainingTags" style="padding:0px;" onchange="selectTag(this.value);">';
+				$html.= '<option value="_add_" selected="selected" disabled="disabled">'.__('add').'...</option>';
+				if (!empty($deselectedTags)) foreach ($deselectedTags as $id => $tag) {
+					$html.= '<option id="tagRem_'.$id.'" value="'.$id.'">'.$tag['label'].'</option>';
```
Comment (matthijskooijman) [inline system/core/tags.php]: Again, `htmlspecialchars()`.

File: system/core/tags.php [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+		$language = hypha_pageGetLanguage($pageListNode, $lang);
+		$url = $language->getAttribute('id').'/'.$language->getAttribute('name');
+		
+		ob_start();
+?>
+<script>
+	function selectTag(id) {
+		if (!id) addTag();
+		else {
+			hypha('<?=$url?>', 'tagSelect', id, function(response) {document.getElementById('tagList').innerHTML = response;});
```
Comment (matthijskooijman) [inline system/core/tags.php]: Again, I would use `json_encode` and `htmlspecialchars` here and in similar places below.

File: system/core/tags.php [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+			document.getElementById('tagList').innerHTML = 'loading...';
+		}
+	}
+	function deselectTag(id) {
+		hypha('<?=$url?>', 'tagDeselect', id, function(response) {document.getElementById('tagList').innerHTML = response;});
+		document.getElementById('tagList').innerHTML = 'loading...';
+	}
+	function addTag() {
+		var name = prompt('<?=__('enter-label')?>', '');
```
Comment (matthijskooijman) [inline system/core/tags.php]: I would use json_encode (and for good measure, also htmlspecialchars though we often leave that out for translations) here and in similar places below.

File: system/core/tags.php [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+	registerCommandCallback('tagNew', 'tagNewCallback');
+	
+	function tagSelectCallback($id) {
+		global $hyphaXml;
+		global $O_O;
+		
+		$lang = $O_O->getContentLanguage();
+		
+		$hyphaXml->lockAndReload();
+		$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
```
Comment (matthijskooijman) [inline system/core/tags.php]: I'm not so sure I like this, since it partially redoes request routing (url to page). The existince of `getPageName()` makes this a lot better than it would be otherwise, though (and I'm not sure if there's a better and also easy way for this...).

File: system/core/tags.php [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+		$newtag = $hyphaXml->createElement('tag');
+		$newtag->setAttribute('id', $id);
+		$pageListNode->appendChild($newtag);
+		$hyphaXml->saveAndUnlock();
+		
+		echo tagList($pageListNode, $lang);
+		exit;
```
Comment (matthijskooijman) [inline system/core/tags.php]: I don't like an exit here (since it prevents any cleanup or other processing further up the call stack. Returning some AjaxResponse value might be better, but we don't support that (yet). Might be something to keep in mind for later.

File: system/core/tags.php [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+		}
+		
+		// Create new tag, add to tagList and page in hypha.xml
+		if (!$tagExists) {
+			$hyphaXml->lockAndReload();
+			$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
+			
+			$tagList = $hyphaXml->getElementsByTagName('tagList')->Item(0);
+			$newTag = $hyphaXml->createElement('tag');
+			$newTag->setAttribute('id', $newTagId);
```
Comment (matthijskooijman) [inline system/core/tags.php]: This should use `$newTag->generateId()` (defined in database.php) rather than generating an id manually. This also means to use the `xml:id` attribute rather than `id` (which allows using getElementsById, IIRC).

File: system/core/tags.php [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+			$newTagLanguage = $hyphaXml->createElement('language', $newTagDescription);
+			$newTagLanguage->setAttribute('id', $lang);
+			$newTagLanguage->setAttribute('label', $newTagName);
+			$newTag->appendChild($newTagLanguage);
+			$tagList->appendChild($newTag);
+			
+			$newTag = $hyphaXml->createElement('tag');
+			$newTag->setAttribute('id', $newTagId);
```
Comment (matthijskooijman) [inline system/core/tags.php]: This probably *should* still use `id` (or even something else) rather than `xml:id`, since the latter should be unique within the XML document.

File: system/core/events.php (hunk 4 above)
Comment (giplt) [reply system/core/events.php]: Not sure. In both cases an action is triggered on the client (probably by a user clicking something), and the some data gets routed to a specific piece of code on the server. The only difference is whether the outcome is a fresh page render (by means of a reload) or dynamically handling the outcome in the page that already exists on the client (through ajax). So although the behaviour is somewhat different the purpose (following up on client events) is pretty much the same. From that perspective introducing a separate command adds just as much confusion as it resolves. Maybe up for discussion another time, for now I would argue to adopt this approach. Will make it a separate commit though.

File: system/core/tags.php (hunk 5 above)
Comment (giplt) [reply system/core/tags.php]: Kinda intended. Using children would add the necessity to check if the childnode is has 'tag' as tagname. This could be done but adds little in terms of performance or robustness. Lacking a getChildrenByTagName function I chose this approach.

Comment (matthijskooijman) [reply system/core/tags.php]: This could be easily done using `find` and an xpath expression, I think.

File: system/core/tags.php (hunk 13 above)
Comment (matthijskooijman) [reply system/core/tags.php]: We just discussed this and the newtext refactoring will probably significantly change the way commands are handled anyway, so we should probably just leave this as it is now and include this in the newtext design instead.

File: system/core/tags.php (hunk 14 above)
Comment (matthijskooijman) [reply system/core/tags.php]: I've added a TODO for later.

File: system/core/tags.php (hunk 5 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Laurens fixed this with XPath.

File: system/core/tags.php (hunk 6 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Laurens implemented both suggestions.

File: system/core/tags.php (hunk 7 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Fixed by Laurens.

File: system/core/tags.php (hunk 9 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Fixed by Laurens.

Comment (matthijskooijman) [reply system/core/tags.php]: Fixed.

File: system/core/tags.php (hunk 8 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Fixed.

File: system/core/tags.php (hunk 1 above)
Comment (matthijskooijman) [reply system/core/tags.php]: I moved all inline style to the CSS in the theme. I also changed a few "id" attributes to "class", since "id" must be unique within a page, and it does not seem unreasonable for multiple taglists to appear in one HTML page.

I also fixed the escaping.

File: system/core/tags.php (hunk 10 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Fixed.

File: system/core/tags.php (hunk 11 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Fixed.

File: system/core/tags.php (hunk 12 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Fixed.

File: system/core/tags.php (hunk 15 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Fixed.

File: system/core/tags.php (hunk 16 above)
Comment (matthijskooijman) [reply system/core/tags.php]: Let's discuss this separately.

File: system/languages/en.php (hunk 2 above)
Comment (matthijskooijman) [reply system/languages/en.php]: Do we want to use the word "tag" at all? Or are we going to call these "labels" everywhere anyway (like Github does)?

One alternative could be "Enter label for this tag" (which is what technically happens under the hood - a new tag object is created, which is assigned a label attribute).

File: system/core/events.php (hunk 4 above)
Comment (matthijskooijman) [reply system/core/events.php]: Ok, let's leave it like this for now. We might need to revisit this as part of the newtext refactor anyway, since form handling and command routing is changed there anyway.

File: system/core/events.php (hunk 3 above)
Comment (matthijskooijman) [reply system/core/events.php]: I've moved this, and the addition of the callback parameter to `hypha()`, into seperate commits.

File: system/core/tags.php (hunk 11 above)
Comment (matthijskooijman) [reply system/core/tags.php]: I ended up not using `htmlspecialchars` here, since HTML entities are not interpreted inside `<script>` tags. This does leave a small window for problems - if a string value contains `</script>` (or even `</` perhaps), then this could end this script tag. This seems unlikely to happen in practice, though it might be a small security risk. For now, let's just leave it like this (this could perhaps be handled more thoroughly by generating HTML through DomDocument everywhere later).

**Comment by matthijskooijman** [comment]
Laurens and me have done a pass over the code, amending the code. I've force-pushed the result. This resolves all of my remarks. As for my remarks about *where* the code should live and how to structure it, I would suggest postponing that discussion and leaving things as-is now (since the current structure with global functions fits with the rest of Hypha fine).

I think there are a few open point:
 - The comment from Diana above about terminology (Label vs tag).
 - I've changed the code to use `generateId()` and `xml:id`, which is the proper way to define unique ids. However, in some other places, the `id` attribute is still used to *reference* ids (e.g. when referencing a tag id from the pagelist). These cannot be converted to `xml:id` (since they are not unique), but mixing `id` and `xml:id` like this might also be confusing. Perhaps they should be changed to `idref` or just `ref` or something like that? Or more specific `tagid` or `tagref` (though `<tag tagid="..."/>` might also look weird). This might need some face2face discussion.
 - The CSS/layout might need a review, I've not looked at that at all.
 - Any existing customized themes need to be modified to include the `tagList` `div`. Is it time already for some auto-upgrade code for this kind of thing? If so, should it also add CSS automatically?

**Comment by laurensmartina** [comment]
Removed unused `$newTagId` and forced pushed the change.

File: system/core/tags.php [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+		$selectedTagIds = [];
+
+		/** @var HyphaDomElement $tagEl */
+		// Build list of active tags for given page
+		foreach ($pageListNode->findXPath('tag') as $tagEl) $selectedTagIds[] = $tagEl->getAttribute('id');
+
+		$selectedTags = [];
+		$deselectedTags = [];
+
+		$lang = htmlspecialchars($lang);
```
Comment (matthijskooijman) [inline system/core/tags.php]: This seems like the wrong place to encode `$lang`. Ususally, you want to encode things as late as possible, just before inserting into (in this case) HTML. Here, `$lang` is encoded to early, and you can see it is then used in an xpath expression (where it should *not* be HTML-encoded, just Xpath-encoded). In fact, it seems that `$lang` is *only* used as part of xpath, never inserted in HTML, so I think this line can simply be dropped?

File: system/core/tags.php [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+				$html.= '<span title="'.htmlspecialchars($tag['description']).'">'.htmlspecialchars($tag['label']).'</span>';
+				if (isUser()) $html.= ' <span class="delete-tag" title="'.__('remove-tag').'" onclick="deselectTag('.htmlspecialchars(json_encode($id)).');">⨯</span>';
+				$html.= '</div>';
+			}
+			$html.= '</div>';
+			if ((isUser() && !empty($deselectedTags)) || isAdmin()) {
+				$html.= '<select class="remainingTags" onchange="selectTag(this.value);">';
+				$html.= '<option value="_add_" selected="selected" disabled="disabled">'.__('add').'...</option>';
+				foreach ($deselectedTags as $id => $tag) {
+					$html.= '<option id="tagRem_'.htmlspecialchars($id).'" value="'.htmlspecialchars($id).'">'.htmlspecialchars($tag['label']).'</option>';
```
Comment (matthijskooijman) [inline system/core/tags.php]: What does "Rem" mean in the id here? I would think "Remove", but that's not what this select is for? Remaining then?

File: system/core/tags.php [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+		$language = hypha_pageGetLanguage($pageListNode, $lang);
+		$url = $language->getAttribute('id').'/'.$language->getAttribute('name');
+
+		ob_start();
+?>
+<script>
+	function selectTag(id) {
+		if (!id) addTag();
+		else {
+			hypha(<?=json_encode($url)?>, 'tagSelect', id, document.getElementById('<?=$formId?>'), function(response) {document.getElementById('tagList').innerHTML = response;});
```
Comment (matthijskooijman) [inline system/core/tags.php]: I would use `json_encode` for `$formId` here (and below).

File: system/core/tags.php [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+	function deselectTag(id) {
+		hypha(<?=json_encode($url)?>, 'tagDeselect', id, document.getElementById('<?=$formId?>'), function(response) {document.getElementById('tagList').innerHTML = response;});
+		document.getElementById('tagList').innerHTML = 'loading...';
+	}
+	function addTag() {
+		var name = prompt(<?=json_encode(__('enter-label'))?>, '');
+		if (name) {
+			var description = prompt(<?=json_encode(__('enter-description'))?>, '');
+			if (description) {
+				var json = '{ "name" : "' + name + '", "description" : "' + description + '"}';
```
Comment (matthijskooijman) [inline system/core/tags.php]: This breaks encoding when name or descriptoin contains a double-quote. I would instead build a JS object and use [JSON.Stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) to encode that to JSON (alternatively, pass the object directly to the `hypha()` function and let that automatically JSON-encode any non-strings or objects it receives).

File: system/core/tags.php [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+		// Check if tag exists
+		$path = 'hypha/tagList/tag/language[@id=' . xpath_encode($lang) . ' and @label=' . xpath_encode($json['name']) . ']';
+		$tagExists = $hyphaXml->findXPath($path)->count() > 0;
+
+		// Create new tag, add to tagList and page in hypha.xml
+		if (!$tagExists) {
+			$newTagName = $json['name'];
+			$newTagDescription = $json['description'];
+
+			$hyphaXml->lockAndReload();
```
Comment (matthijskooijman) [inline system/core/tags.php]: This lock should probably happen before checking whether the tag exists (probably even before checking whether the tag list exists), since otherwise there is a race condition where the tag list or tag might be added twice if two requests happen at the same time. This does mean the file is locked for a bit longer, but that should be fine.

File: system/core/tags.php [hunk 22]
```diff
@@ -0,0 +1,192 @@
+<?php
+	// TODO:
+	// - add styling for default theme
```
Comment (matthijskooijman) [inline system/core/tags.php]: I believe this TODO is resolved with Tammo's commits? The rest below will probably stay until after this first version is merged.

**Comment by matthijskooijman** [comment]
@tammoterhark, there are still quite some open comments that would need to be resolved (or we should decide to not fix them or fix them later, but then we should actually decide that).

File: system/languages/en.php (hunk 2 above)
Comment (matthijskooijman) [reply system/languages/en.php]: Conclusion: Let's use "Tag" everywhere, also in the Dutch translation and string ids.

**Comment by matthijskooijman** [comment]
> When adding a tag with an obvious title (example: "Amersfoort" ) the tagging system still needs a description. If you do not supply a description, the tag is not created, no user feedback given.

Conclusion: Make description required, if it is left empty, just ask it again (maybe add a message saying it is required).

**Comment by tammoterhark** [comment]
Datatype shown as datatype.name.tag_list which is possibly unclear for user.

Would prefer: tagpage (since taglist would be a list of tags)

Firefox_Screenshot_2020-02-24T12-07-31 342Z

**Comment by matthijskooijman** [comment]
I've done some more minor fixes, as well as split some commits. @tammoterhark, you had commented the CSS rule for `.prefix` (probably to prevent it from interfering with the taglist), but that changes the languagelist. Instead, I made the `.prefix` rule more specific: b88eb77e7fa41cc195d63e80ea1ed930613bcc11. I also split off some CSS / HTML changes not necessarily to the taglist: 8d8c758b6a19f5398f2912b1009516bc5fd403be and 60d3116be20a43e3c5d03e8a26f1d7fc6363a183. In the future, it would be easier if such unrelated changes would be made in a separate commit directly.

I've also changed terminology to call the tags "tags" rather than label. Note that the terminology "label" and "tag" are both used: A tag *has* a label (and a description). In practice, the label and the tag are usually considered the same thing.

**Comment by matthijskooijman** [comment]
@tammoterhark, your last remark is out of scope for this PR, but I expect that @laurensmartina will solve it before creating a PR for the "taglist" (or what it's called) branch.
